### DISCUSSION TITLE
Webサーバー経由の/growi/のようなサブディレクトリで運用する方法

### DISCUSSION THREAD

**Discussion by rsogo** [growilabs/growi#10424] [Help / Questions] [answered]
Environment
------------

### Host

| item     | version |
| ---      | --- |
|GROWI     | 3.2.10-RC |
|node.js   | 8.11.1 |
|npm       | 6.3.0 |
|Using Docker|no|
|Using [growi-docker-compose][growi-docker-compose]|no|

How to reproduce? (再現手順)
---------------------------

次のような構成をとっています。

Client (Browser) --> Nginx --> Growi

Nginxに以下のように`/growi/`というパスにアクセスすると、
growiにルーティングするように設定。

```
location /growi/ {

      add_header Access-Control-Allow-Origin '*';
      add_header Access-Control-Allow-Methods 'GET, POST, PUT, DELETE';
      add_header Access-Control-Allow-Headers 'Origin, Authorization, Accept, Content-Type';
      add_header Access-Control-Allow-Credentials true;
      proxy_pass http://localhost:3000/;
}
```

What happens? (症状)
---------------------

`http://{nginxのhostname}:{nginxのport}/growi/`

にアクセスすると、各種リソースへのリンクが404になっていたり、
各種操作がエラーになります。

What is the expected result? (期待される動作)
-------------------------------------------

ソースコードを変更して、サブディレクトリ対応をするか検討していますが、
変更箇所が広範囲になりそうでしたので、以下、教えていただけないでしょうか。

- `/growi` のような任意のサブディレクトリで運用されている方はどうされてますか？
- 今後、`/growi` のような任意のサブディレクトリで動作するように対応される予定はありますでしょうか？

背景としては、同一サーバー内に複数のソフトウェアを運用していますが、
共通のnginxをproxyとして経由させることでユーザーからは単一のホストとして公開し、
サブディレクトリを切って、各ソフトウェアにアクセスさせたいと考えています。

**Comment by yuki-takei** [comment, marked as answer]
@rsogo GROWI を開発する前、Crowi を使っていた僕も同じように subdir で運用したいと考えたものでした…

https://github.com/crowi/crowi/issues/162

本家は対応の予定なしということで、fork した後、中身がそこそこ分かってきたら自分でやろうかなとは思っていたのですが、やはり変更箇所が相当多くて「これは無理だなあ」というのが正直な感想でした。


> /growi のような任意のサブディレクトリで運用されている方はどうされてますか？

弊社ではその時期を境に、他のシステムも含め全部サブドメインで分けるやり方で統一する方にシフトしました。
僕個人の美意識的には subdir で統一したかったのですが、結果的には、「全てのシステムを subdir 統一」よりも「全てのシステムをサブドメイン統一」の方が設定コスト的には低くなったと感じています。(そういう問題じゃない、というのはその通り)

改修してやってる人がいたら是非ソースを拝見させていただきたい。


> 今後、/growi のような任意のサブディレクトリで動作するように対応される予定はありますでしょうか？

申し訳ないのですが、上記の通り自社でサブドメインに舵を切ってしまったので、現時点では優先度もモチベーションもかなり低いのが実際の所です…

**Comment by rsogo** [comment]
@yuki-takei ありがとうございます。

一応、どれだけ大変そうかあたりをつけるために、env.prod.jsに`ROOT_DIR: '/growi',`という設定をもたせて、以下のようなパスを指定しているところをチクチク変更していって、大まかな動作はするようになったのですが、細かいところを踏まえるとまだまだ先は長いなという感触です。

```
-      <form role="form" action="/installer/createAdmin" method="post" id="register-form">
+      <form role="form" action={process.env.ROOT_DIR  + "/installer/createAdmin"} method="post" id="register-form">
```

```
-      axios[method](`/_api${path}`, params)
+      axios[method](process.env.ROOT_DIR + "/_api${path}", params)
```

**Comment by rsogo** [comment]
現状、理解できましたのでクローズします。
ありがとうございました。
